Full focal plane infrastructure

welucas2 opened the pull request:
This isn't quite ready yet, but I'd like to get a draft PR in place. This will add to imSim the capability to simulate the full focal plane. Currently, we can only simulate detectors in isolation: for a given set of observation parameters and input, imSim will run through tasks drawing each detector in the list. This neglects any 'crosstalk' between detectors in the form of photons that are generated for one but which (for whatever reason) end up raining down on another.

We're building towards a two-pass full focal plane model which will work like the following:
1) Pass one will work like a current imSim run, with the addition of an extra output field `off_detector_photons`. While working on each detector, imSim will gather the photons that end up outside its bounds and write them to file for use in pass two. The other output is the same image that would normally have been produced, of those objects which lie within the detector's bounds and have been drawn on it.
2) Pass two takes both of those outputs from pass one. While working on a detector, it will read its image from pass one as well as the off-detector photons that were originally created by tasks working on other detectors. In this way, the task working on detector X is able to draw the photons that were originally created while working on the other 188 tasks.

As of the moment there are still things to do here:
* The current implementation is for the photon pooling pipeline with `LSST_PhotonPoolingImage`. I'd like to see how easy it would be to implement for the original `LSST_Image` pipeline, too. This might be something to do later on separately.
  * If not doing now, preventing or providing a warning for users who try to use the `off_detector_photons` extra output with the original pipeline.
* Provide a switch when reading `off_detector_photons` in pass two that refactors the input at the very start, prior to drawing the images.
  * Each task in pass one produces one file. which could potentially contain photons destined for all 188 other detectors. So, to be sure that all photons are captured, *not* refactoring the files would require each task in pass two to read and work with 188 files.
  * Refactoring would take the 189 files, each containing all the photons originating from a given detector but destined elsewhere, and produce a different set of 189 files, each containing the photons from any origin but destined for a given detector. With this, each task in pass two would only need to read the one off-detector photon file.
* We should provide template configuration files for users, and probably expand the current `imsim-user-instcat-ffp-pass1.yaml` and `imsim-user-instcat-ffp-pass2.yaml` examples.

welucas2 committed:
Machinery to write photons outside image bounds as extra output to FITS.

welucas2 committed:
Put scattered photons in focal plane coords.

welucas2 committed:
First partial attempt at reading in the photons from file.

welucas2 committed:
Working scattered photon input

welucas2 committed:
Initial full working pipeline

welucas2 committed:
Full focal plane tests

welucas2 committed:
Move gathering of scattered photons into new function in photon_scattering.py

welucas2 committed:
Initial go at code to find a science detectors neighbours as part of the Camera class.

welucas2 committed:
Make references to 'scattered' photons 'off-detector' photons instead -- more general, as we don't want to presume how they got off the detector. Be a bit more general.

welucas2 committed:
Rename photon_scattering.py to full_focal_plane.py

welucas2 committed:
Changes to variable names in get_adjacent_detectors()

welucas2 committed:
Move off detector photons from photon pooling config to FFP pass 1 example

welucas2 committed:
Doing some horrible initial experimentation in instcat.py.

welucas2 committed:
Make the fiducial WCS a component of the instcat in the YAML config.

welucas2 committed:
Big improvements to the instcat multi detector object handling. And it now works.

welucas2 committed:
First go at getting skyCatalog interface to accept/reject objects by proximity to detector.

welucas2 committed:
Add example_instance_catalog_ffp.txt

welucas2 committed:
Add the fiducial WCS to imsim-config-skycat.yaml.

welucas2 committed:
First attempt at logic to filter out objects closer to other dets in skycat.py

jchiang87 committed:
keep track of objects to skip instead of trying to use an iterator provided by the filter built-in

jchiang87 committed:
include f-string prefix

jchiang87 committed:
don't run AddNoise for pass 2 rendering

welucas2 committed:
Config now keys off-detector photons to det_num/name.

welucas2 committed:
Fix test_trimmer.py to set edge_pix=100, the old default, to keep a test useful. (Default is now 200 pixels to cover the gap between detectors.)

welucas2 committed:
New input.instance_catalog.full_focal_plane enables and disables FFP.

welucas2 committed:
Have test_image_nobjects set edge_pix=100 as the new default of 200 (to fill detector gaps) changes results.

welucas2 committed:
Calculate fiducial WCS before image WCS in order to correctly store _icrf_to_field in base config.

welucas2 committed:
Add full_focal_plane parameter to skycat and ensure correct _icrf_to_field is stored.

welucas2 committed:
Allow off_detector_photons to be read from multiple files.

welucas2 committed:
Update tests to use LsstCamSim.

welucas2 committed:
Fix edge_pix values used in test_stamp.py to old value of 100.

welucas2 committed:
Add tests for off-detector photon loader.

welucas2 committed:
New test for LSST_FocalPlaneImage.

welucas2 committed:
Rework off-detector photons input to lazily read photons and also allow them to be manually set if desired, with what are hopefully appropriate warnings along the way.

welucas2 committed:
Minor tidying up in the second pass example config.

welucas2 committed:
Add files in tests/data for new full focal plane tests.

rmjarvis commented on examples/imsim-user-instcat-ffp-pass2.yaml:
```diff
@@ -0,0 +1,81 @@
+        items:
+            - { type: OpsimData, field: observationId }
+            - { type: OpsimData, field: snap }
+            - $band
```
A maybe simpler API would be to just have `file_name`, but allow `*`.  Then later use `glob.glob(file_name)` to get the list of files matching that pattern.  I use this in Piff when the input can involve multiple files.

rmjarvis commented on imsim/instcat.py:
```diff
@@ -244,8 +274,16 @@ def _load(self, logger=None):
+                        # Check whether the object is closer to this detector's center
+                        # than any of the adjacent detectors. If not, then skip it.
+                        obj_dist = {det: detector_centers[det].distanceTo(world_pos) for det in adjacent_detectors}
+                        closest_detector = min(obj_dist, key=obj_dist.get)
```
This is very nice.  Took me a minute to understand how it worked, but it's quite elegant.

rmjarvis reviewed (commented):
Looks good.  Some suggestions here, but I think all minor edits.

jchiang87 committed:
remove off-detector photons as they are read in for pass2; provide more general gather_photons_by_bounds for detector-level downselection; update tests

jchiang87 committed:
set image.sky_level and output.cosmic_ray_rate to zero for pass2 since they are already added in pass1

also disable output.photon_pooling_truth for pass2 to avoid overwriting centroid files from pass1

jchiang87 commented on imsim/full_focal_plane.py:
```diff
@@ -0,0 +1,238 @@
+        if len(self.file_names) > 0:
+            photon_storage = []
+            logger.info(f"Detector {self.det_name} reading {len(self.file_names)} files for off-detector photons.")
+            for file_name in self.file_names:
```
For a full focal plane simulation, using the diffsky + star catalogs,  I'm finding that the job memory blows up here to at least 40 GBs if we just read in all of the off-detector photons from the other 188 detectors (in my initial test it looks like it OOM'd on a perlmutter login node).  I've made some modifications [here](https://github.com/LSSTDESC/imSim/compare/u/welucas2/full-focal-plane...u/jchiang/ffp_off_det_photon_memory_fixes) to downselect for the CCD being processed.  With this change, the increase in memory is only a few hundred MB in my test.

jchiang87 commented on examples/imsim-user-instcat-ffp-pass2.yaml:
```diff
@@ -0,0 +1,81 @@
+output.nfiles: 1
+output.file_name.format: "eimage_full_focal_%08d-%1d-%s-%s-det%03d.fits"
+
+output.dir: output
```
Since the sky background and cosmic rays are added to the eimage in pass1, I'd add these lines here:
```
image.sky_level: 0
output.cosmic_ray_rate: 0.0
```
I added these to the [branch](https://github.com/LSSTDESC/imSim/compare/u/welucas2/full-focal-plane...u/jchiang/ffp_off_det_photon_memory_fixes) I made off of this one.

Edit:  Something else I just noticed:  the centroid file from pass1 was overwritten with an empty file as an outcome of pass2.  So I would also add
```
output.photon_pooling_truth: ""
```
to avoid that.

jchiang87 reviewed (commented):
Here are some observations and suggested code updates based on my running a full focal plane simulation at NERSC using full diffsky + stars skyCatalogs.

One other thing I noticed after running pass2 on my test CCD (R22_S11):  Objects near the edge of the CCD (but definitely on that CCD) got brighter after pass2 vs pass1, which suggests double counting of those objects.  I can search through the centroid files to see which objects were simulated in pass1 for each CCD, but I'm wondering if it's possible to include object ID information in the photon arrays that are written out to the off-detector files.   That would be useful if we find that we need to filter those photons in pass2.

I confirm that I see the same objects in the pass1 centroid files for adjacent CCDs at the ~5-7% level.  For one pair, I see ~3000 objects in common out of 44k and 48k objects in the two files, respectively.  (This seems to have occurred because I didn't set `input.sky_catalog.full_focal_plane: True` in my pass1 yaml file.  I've noted below that an entry like this should be in the instcat-based yaml example.)

I'm seeing `nan`s for the `x` and `y` values in the off-detector photon files.  It looks like there are just a handful of photons per file (~10), but they are in every file.

jchiang87 commented on examples/imsim-user-instcat-ffp-pass1.yaml:
```diff
@@ -0,0 +1,56 @@
+
+input.instance_catalog.file_name: $os.environ.get('IMSIM_HOME')+'/imSim/examples/example_instance_catalog_ffp.txt'
+input.instance_catalog.sort_mag: False
+
```
There needs to be an entry
```
input.instance_catalog.full_focal_plane: True
```
here. The same goes for a skycatalogs-based yaml file:
```
input.sky_catalog.full_focal_plane: True
```
Otherwise, the code in the relevant catalog implementations for calculating computing the closest detector won't be executed, and we'll get multiple renderings of objects at the CCD edges.

jchiang87 commented on imsim/skycat.py:
```diff
@@ -217,13 +273,30 @@ def getKwargs(self, config, base, logger):
                'approx_nobjects': int,
                'mjd': float,
                'pupil_area': float,
+               'full_focal_plane': bool,
```
If we specify `output.off_detector_photons` in the yaml config, then I think `full_focal_plane` should be `True`.  Conversely, if we don't do the `full_focal_plane` calculation in the skycat or instcat code, then setting `output.off_detector_photons` in the yaml isn't useful.  So I'd omit `full_focal_plane` from `opt` here, and instead infer the value from the config, e.g., adding code like this below:
```
        try:
            galsim.config.GetFromConfig(base, 'output.off_detector_photons')
            kwargs['full_focal_plane'] = True
        except galsim.config.GalSimConfigError:
            kwargs['full_focal_plane'] = False
```

rmjarvis replied:
I'd prefer full_focal plane remain in the config as an explicit specification, but have it be an error if the value of full_focal_plane is incompatible with the presence or not of off_detector_photons.

jchiang87 committed:
Update imsim/skycat.py

Co-authored-by: Mike Jarvis <michael@jarvis.net>

jchiang87 committed:
use a template based on image.wcs for the fiducial_wcs

jchiang87 committed:
use a glob pattern to find off-detector photon files

jchiang87 committed:
ensure there is only one photon_array in self.final_data

jchiang87 committed:
raise an exception of .writeHdu is called

jchiang87 committed:
use GetInputObj instead of accessing config['_input_objs'] directly

jchiang87 committed:
set input.instance_catalog.full_focal_plane to True; disable checkpointing and output.readout

jchiang87 committed:
remove the .photons setter

jchiang87 committed:
remove non-linear distortions from fiducial wcs

jchiang87 committed:
add check that full_focal_plane for instcat and skycat is set consistently with output.off_detector_photons

jchiang87 commented:
@rmjarvis I think I've largely addressed your comments as well as the items I brought up.  (There's some consolidation of the common code in `instcat.py` and `skycat.py` that I'd like to defer to a different PR.)  Please have a look and let me know if there are any remaining issues.

rmjarvis commented on imsim/full_focal_plane.py:
```diff
@@ -128,52 +131,29 @@ def __init__(self, camera, det_name, file_name=None, file_pattern=None, photons=
-                          ]
+            if dir is not None:
+                file_name = os.path.join(dir, file_name)
+            self.file_names = sorted(fname for fname in glob.glob(file_name))
```
I would have thought simply `sorted(glob.glob(file_name))` would suffice here.

rmjarvis reviewed (approved):
Sorry to take so long on this.  I just have a slight simplification suggestions.  The code looks fine to me.

jchiang87 committed:
simplify the argument to sorted
